app/src/main/java/com/example/cinemapp/ui/main/HomeViewModel.kt

```diff
@@ -0,0 +1,54 @@
+package com.example.cinemapp.ui.main
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.cinemapp.data.MovieCard
+import com.example.cinemapp.data.MovieRepository
+import com.example.cinemapp.util.MovieUtil
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+class HomeViewModel(
+    private val movieRepository: MovieRepository
+) : ViewModel() {
+
+    data class State(
+        val movies: List<MovieCard> = emptyList(),
+        val search: String = "",
+        val isTimeToPaginate: Boolean = false,
+        val pagesLoaded: Int = 0,
+    )
+
+    private val _state = MutableStateFlow(State())
+    val state: StateFlow<State> = _state
+
+    fun getUpcoming() {
+        viewModelScope.launch {
+            _state.update {
+                it.copy(movies = movieRepository.getUpcoming()
+                    ?.let { list -> MovieUtil.map(list) } ?: emptyList())
+            }
+        }
+    }
+
+    fun getUpcomingNextPage() {
+        setTimeToPaginate(true)
```

Comment:
This seems a little bit confusing 😅 Here, you set `isTimeToPaginate` to `true`, but then later, upon updating state again, you set it to false...Seems like `isTimeToPaginate` should be a fire-and-forget event, right? If so, you should make create a new `SharedFlow` variable in ViewModel, collect its value in fragment, and act accordingly.

Reply:
This is so that only one call of this function will be done as the function that checks if user is near the end of the Recycle view would be called repeatedly otherwise.